articles/content_filtering.md

```diff
@@ -0,0 +1,315 @@
+
+### Action Topics
+
+Each action server will provide two topics named `feedback` and `status` as followings,
+
+<img src="../img/content_filter/action_architecture_overview.png" width="600">
+
+`feedback` and `status` are topics mapped with action name and published by action server, and action clients subscribe those topics to get feedback and status.
+When there are many goals from many clients, the choice to have a single `feedback` and `status` topic per action server is suboptimal in terms of processing and bandwidth resource. 
+It is up to clients to filter out feedback/status messages that are not pertinent to them.
+In this scenario, M goals are sent to N clients there is an unnecessary use of bandwidth and processing.
+Especially in extreme cases where M and N are large. (each goal is identified by goal ID based on uuid, filtering out the goal id is done by client library so user application does not need to care.)
+
+## ContentFilteredTopic
+
+ContentFilteredTopic describes a more sophisticated subscription that indicates the subscriber does not want to necessarily see all values of each instance published under the Topic.
+Rather, it wants to see only the values whose contents satisfy certain criteria.
+This class therefore can be used to request content-based subscriptions. 
+
+### Specification
+
+- Multiple ContentFilteredTopicDataReader exist and filtering is dependent on each ContentFilteredTopicDataReader's logical filtering expression and parameters.
+- Subscriber is responsible to create ContentFilteredTopic.
+- ContentFilteredTopicDataWriter does filtering for each ContentFilteredTopicDataReader endpoint.
+  Basically it is likely that ContentFilteredTopicDataWriter does filtering on writer side, so that it can provide the improvement and efficiency.
+  But it is dependent on imnplementation if always ContentFilteredTopicDataWriter does filtering.
+  This is because if there is a lot of ContentFilteredTopicDataReader with complicated filtering expression and parameter, it would be probably more efficient just to send the messages to all ContentFilteredTopicDataReader.
+  And ContentFilteredTopicDataReader can filter out the message if not necessary based on filtering expression and parameter.
+  This is also said that ContentFilteredTopicDataReader does need to do filtering always just in case ContentFilteredTopicDataWriter ends up sending w/o filitering.
+  <img src="../img/content_filter/contents_filter_overview.png" width="600">
+- The filtering is done by means of evaluating a logical expression that involves the values some of the data-fields in the sample.
+  The logical expression is derived from the filter_expression and expression_parameters arguments.
+- Multiple filter_expression and expression_parameters can be supported on single ContentFilteredTopic.

```

Comment:
Each ContentFilteredTopic has a single filter_expression and a sequence of expression_parameters

Reply:
thanks, i borrowed your sentence here.